In [21]:
import pandas as pd
import numpy as np
from datetime import datetime as dt
import matplotlib.pyplot  as plt

# Purpose

Author: Chris Kuzemka

The purpose of this notebook is to check what is housed inside the kaggle dataset of flight itineraries. The dataset is a large and computationally expensive. It can be downloaded [here](https://www.kaggle.com/datasets/dilwong/flightprices/data). 

Details and a report going over what is found will be placed here as well as a summary of what to clean and change. 

# Data

In [22]:
# do this to show all the columns....note setting None means there are no restrictions on any of this. 
pd.set_option("display.max_columns", None)
pd.set_option('display.max_rows', 300)

In [23]:
df = pd.read_csv(r"C:\Users\User\OneDrive\Desktop\Fordham\Programming with Python (CISC-5380-L01)\intro_to_python_project\data\jetblue_df.csv")

In [24]:
df.head()

Unnamed: 0                             legId  searchDate  flightDate  \
0           5  79eda9f841e226a1e2121d74211e595c  2022-04-16  2022-04-17   
1          29  24c07cf38de8a7a0ed2d7709168042a8  2022-04-16  2022-04-17   
2          57  a291a2b5ac19223c3407cfec58d1fc24  2022-04-16  2022-04-17   
3         146  d53bcf9e8b6523fa2de9420d7cb22196  2022-04-16  2022-04-17   
4         209  5bb138f39804320ece767e747608a50e  2022-04-16  2022-04-17   

  startingAirport destinationAirport fareBasisCode travelDuration  \
0             ATL                BOS      VH0AUEL1        PT2H38M   
1             ATL                BOS      QH0KUEL1        PT2H44M   
2             ATL                CLT      VH0AUEL1        PT8H48M   
3             ATL                DTW      RH0AUEL1         PT5H8M   
4             ATL                EWR      VC0QBEL1         PT2H9M   

   elapsedDays  isBasicEconomy  isRefundable  isNonStop  baseFare  totalFare  \
0            0           False         False       True    217.67     248.60   
1            0           False         False       True    316.73     355.09   
2            0           False         False      False    482.79     542.60   
3            0           False         False      False    269.78     313.61   
4            0           False         False       True    171.16     198.60   

   seatsRemaining  totalTravelDistance segmentsDepartureTimeEpochSeconds  \
0               7                947.0                        1650206700   
1               7                947.0                        1650233220   
2               7               1675.0            1650206700||1650229500   
3               7               1247.0            1650235800||1650246600   
4               7                762.0                        1650195000   

                            segmentsDepartureTimeRaw  \
0                      2022-04-17T10:45:00.000-04:00   
1                      2022-04-17T18:07:00.000-04:00   
2  2022-04-17T10:45:00.000-04:00||2022-04-17T17:0...   
3  2022-04-17T18:50:00.000-04:00||2022-04-17T21:5...   
4                      2022-04-17T07:30:00.000-04:00   

  segmentsArrivalTimeEpochSeconds  \
0                      1650216180   
1                      1650243060   
2          1650216180||1650238380   
3          1650244380||1650254280   
4                      1650202740   

                              segmentsArrivalTimeRaw  \
0                      2022-04-17T13:23:00.000-04:00   
1                      2022-04-17T20:51:00.000-04:00   
2  2022-04-17T13:23:00.000-04:00||2022-04-17T19:3...   
3  2022-04-17T21:13:00.000-04:00||2022-04-17T23:5...   
4                      2022-04-17T09:39:00.000-04:00   

  segmentsArrivalAirportCode segmentsDepartureAirportCode  \
0                        BOS                          ATL   
1                        BOS                          ATL   
2                   BOS||CLT                     ATL||BOS   
3                   JFK||DTW                     ATL||JFK   
4                        EWR                          ATL   

                segmentsAirlineName segmentsAirlineCode  \
0                   JetBlue Airways                  B6   
1                   JetBlue Airways                  B6   
2  JetBlue Airways||JetBlue Airways              B6||B6   
3  JetBlue Airways||JetBlue Airways              B6||B6   
4                   JetBlue Airways                  B6   

  segmentsEquipmentDescription segmentsDurationInSeconds segmentsDistance  \
0                          NaN                      9480              947   
1                          NaN                      9840              947   
2                ||Embraer 190                9480||8880         947||728   
3     Embraer 190||Embraer 190                8580||7680         762||485   
4                  Embraer 190                      7740              762   

  segmentsCabinCode  
0             coach  
1             coach  
2      coach||coach  
3      coach||coach  
4      

In [25]:
sorted(df.columns)

['Unnamed: 0',
 'baseFare',
 'destinationAirport',
 'elapsedDays',
 'fareBasisCode',
 'flightDate',
 'isBasicEconomy',
 'isNonStop',
 'isRefundable',
 'legId',
 'searchDate',
 'seatsRemaining',
 'segmentsAirlineCode',
 'segmentsAirlineName',
 'segmentsArrivalAirportCode',
 'segmentsArrivalTimeEpochSeconds',
 'segmentsArrivalTimeRaw',
 'segmentsCabinCode',
 'segmentsDepartureAirportCode',
 'segmentsDepartureTimeEpochSeconds',
 'segmentsDepartureTimeRaw',
 'segmentsDistance',
 'segmentsDurationInSeconds',
 'segmentsEquipmentDescription',
 'startingAirport',
 'totalFare',
 'totalTravelDistance',
 'travelDuration']

In [26]:
df.shape

(6824440, 28)

In [27]:
df.size

191084320

# EDA

## Null Handling

In [28]:
print("Finding nulls per column")
df.isnull().any()

Finding nulls per column


Unnamed: 0                           False
legId                                False
searchDate                           False
flightDate                           False
startingAirport                      False
destinationAirport                   False
fareBasisCode                        False
travelDuration                       False
elapsedDays                          False
isBasicEconomy                       False
isRefundable                         False
isNonStop                            False
baseFare                             False
totalFare                            False
seatsRemaining                       False
totalTravelDistance                   True
segmentsDepartureTimeEpochSeconds    False
segmentsDepartureTimeRaw             False
segmentsArrivalTimeEpochSeconds      False
segmentsArrivalTimeRaw               False
segmentsArrivalAirportCode           False
segmentsDepartureAirportCode         False
segmentsAirlineName                  False
segmentsAir

Above, we find that `totalTravelDistance`, `segmentsEquipmentDescription`, and `segmentsDistance` contain nulls. Now, let's see the information on these columns in particular. 

In [29]:
print("Observing null columns")
df[["totalTravelDistance", "segmentsEquipmentDescription", "segmentsDistance"]].head()

Observing null columns


totalTravelDistance segmentsEquipmentDescription segmentsDistance
0                947.0                          NaN              947
1                947.0                          NaN              947
2               1675.0                ||Embraer 190         947||728
3               1247.0     Embraer 190||Embraer 190         762||485
4                762.0                  Embraer 190              762

<a id='questions_1'></a>

The information above showcases distances and types of aircraft. This information could be important in building a model that can use different features to ultimately predict price. However, some of this information may also contribute nothing more than bloat to a model. Some questions that come to mind:

- <font color="green">[CLOSED]</font> what is the difference between `totalTravelDistance` and `segmentsDistance`?
    -  per the dictionary: `totalTravelDistance` is the total distance traveled in miles. It's reported by the dataset author that some of this data may be missing. The `segmentsDistance` is the distance in miles for each leg of the trip. 
- <font color="red">[OPEN]</font> will an end-product tremendously benefit from the knowledge of the kind of aircraft found in `segmentsEquipmentDescription`?
    - not truly sure when we can determine if this will be useful or not for a model. 

In [30]:
def detect_nulls(df):
    null_counts = df.isna().sum()
    null_percentages = (null_counts / len(df)) * 100
    null_summary = pd.concat([null_counts, null_percentages], axis=1, keys=['Null Count', 'Null Percentage'])
    null_summary = null_summary[null_summary['Null Count']>0]
    return null_summary

# Detect and display nulls
null_summary = detect_nulls(df)
display(null_summary)
print("\nTotal number of rows:", len(df))

Null Count  Null Percentage
totalTravelDistance                56436         0.826969
segmentsEquipmentDescription       52443         0.768459
segmentsDistance                   19469         0.285283


Total number of rows: 6824440


The biggest offender of nulls is the `totalTravelDistance` column, where ~0.83% of this column contains nulls.

Here is when we looked at the data dictionary and were able to address more information on this column. Thus, we were able to also simultaneously address the first question found in [the previous cell housing some initial questions](#questions_1)

Per the linked cell written in this current cell, it is expected that this column contains missing values. However, we do not know why. 

The next step will focus on some of these instances where `totalTravelDistance` is null. Furthermore, we should note if `segmentsDistance` is also null in these areas. There may be a possibility where the segments feature might have the information missing in the total distance feature or vice versa. Or they both will be null together where appropriate. 

In [31]:
 # Investigating relationships between null values

null_combinations = df.groupby([
    df['totalTravelDistance'].isnull(),
    df['segmentsEquipmentDescription'].isnull(),
    df['segmentsDistance'].isnull()
]).size()
display(null_combinations)

totalTravelDistance  segmentsEquipmentDescription  segmentsDistance
False                False                         False               6715947
                     True                          False                 52057
True                 False                         False                 36967
                                                   True                  19083
                     True                          True                    386
dtype: int64

1. Rows with no nulls: 6715947 (98.41%)
2. Rows with null only in segmentsEquipmentDescription: 52057 (0.76%)
3. Rows with null only in totalTravelDistance: 36967 (0.54%)
4. Rows with null in totalTravelDistance and segmentsDistance: 19083 (0.28%)
5. Rows with null in all three columns: 386 (0.01%)

The dataset has a very low percentage of null values overall.
Most rows (98.41%) have no nulls in these columns.
The null percentages for each column are still below 1%.

In [32]:
df.dropna(inplace=True)
df.shape

(6715947, 28)

I saw some places where `"||"` exists for either segments distance or equipment. 

To see such, I should look at the distinct values of these dataframes. 

In [33]:
columns_to_see_classification_error_in = ["segmentsEquipmentDescription", "segmentsAirlineCode", "segmentsAirlineName", "segmentsArrivalAirportCode", "segmentsArrivalTimeEpochSeconds", "segmentsDepartureTimeEpochSeconds", "segmentsCabinCode"]
print(columns_to_see_classification_error_in)
for col in columns_to_see_classification_error_in[1:3]:
    print(f'unique value without duplicates for column {col}')
    print(sorted(df[col].unique()))


['segmentsEquipmentDescription', 'segmentsAirlineCode', 'segmentsAirlineName', 'segmentsArrivalAirportCode', 'segmentsArrivalTimeEpochSeconds', 'segmentsDepartureTimeEpochSeconds', 'segmentsCabinCode']
unique value without duplicates for column segmentsAirlineCode
['9K||9K||9K||B6', '9K||9K||B6', '9K||B6', '9K||B6||B6', 'AA||B6', 'B6', 'B6||3M||B6', 'B6||9K', 'B6||9K||9K', 'B6||9K||B6', 'B6||AA', 'B6||B6', 'B6||B6||9K', 'B6||B6||B6']
unique value without duplicates for column segmentsAirlineName
['American Airlines||JetBlue Airways', 'Cape Air||Cape Air||Cape Air||JetBlue Airways', 'Cape Air||Cape Air||JetBlue Airways', 'Cape Air||JetBlue Airways', 'Cape Air||JetBlue Airways||JetBlue Airways', 'JetBlue Airways', 'JetBlue Airways||American Airlines', 'JetBlue Airways||Cape Air', 'JetBlue Airways||Cape Air||Cape Air', 'JetBlue Airways||Cape Air||JetBlue Airways', 'JetBlue Airways||JetBlue Airways', 'JetBlue Airways||JetBlue Airways||Cape Air', 'JetBlue Airways||JetBlue Airways||JetBlue A

Consider dropping `segmentsAirlineCode` and `segments EquimentDescription`

In [34]:
df.drop(columns=['segmentsEquipmentDescription','segmentsAirlineCode','Unnamed: 0'],inplace=True)
sorted(df.columns)

['baseFare',
 'destinationAirport',
 'elapsedDays',
 'fareBasisCode',
 'flightDate',
 'isBasicEconomy',
 'isNonStop',
 'isRefundable',
 'legId',
 'searchDate',
 'seatsRemaining',
 'segmentsAirlineName',
 'segmentsArrivalAirportCode',
 'segmentsArrivalTimeEpochSeconds',
 'segmentsArrivalTimeRaw',
 'segmentsCabinCode',
 'segmentsDepartureAirportCode',
 'segmentsDepartureTimeEpochSeconds',
 'segmentsDepartureTimeRaw',
 'segmentsDistance',
 'segmentsDurationInSeconds',
 'startingAirport',
 'totalFare',
 'totalTravelDistance',
 'travelDuration']

In [35]:
df.shape

(6715947, 25)

In [36]:
df.groupby(df['segmentsAirlineName']).size().sort_values(ascending=False)

segmentsAirlineName
JetBlue Airways                                      3506867
JetBlue Airways||JetBlue Airways                     3092858
JetBlue Airways||JetBlue Airways||JetBlue Airways     108877
JetBlue Airways||Cape Air||Cape Air                     2142
Cape Air||JetBlue Airways                               1842
JetBlue Airways||Cape Air                               1353
JetBlue Airways||Cape Air||JetBlue Airways              1035
American Airlines||JetBlue Airways                       388
Cape Air||Cape Air||JetBlue Airways                      249
JetBlue Airways||JetBlue Airways||Cape Air               126
Cape Air||JetBlue Airways||JetBlue Airways               116
JetBlue Airways||Silver Airways||JetBlue Airways          80
JetBlue Airways||American Airlines                        13
Cape Air||Cape Air||Cape Air||JetBlue Airways              1
dtype: int64

In [37]:
df.head()

legId  searchDate  flightDate startingAirport  \
2  a291a2b5ac19223c3407cfec58d1fc24  2022-04-16  2022-04-17             ATL   
3  d53bcf9e8b6523fa2de9420d7cb22196  2022-04-16  2022-04-17             ATL   
4  5bb138f39804320ece767e747608a50e  2022-04-16  2022-04-17             ATL   
5  8201f75d221c7aa5b78f3b7781166032  2022-04-16  2022-04-17             ATL   
6  a63591f2b4df39d809c4a45a3898e122  2022-04-16  2022-04-17             ATL   

  destinationAirport fareBasisCode travelDuration  elapsedDays  \
2                CLT      VH0AUEL1        PT8H48M            0   
3                DTW      RH0AUEL1         PT5H8M            0   
4                EWR      VC0QBEL1         PT2H9M            0   
5                JFK      RC0ABEL1        PT2H15M            0   
6                JFK      RC0ABEL1        PT2H23M            0   

   isBasicEconomy  isRefundable  isNonStop  baseFare  totalFare  \
2           False         False      False    482.79     542.60   
3           False         False      False    269.78     313.61   
4           False         False       True    171.16     198.60   
5           False         False       True    171.16     198.60   
6           False         False       True    171.16     198.60   

   seatsRemaining  totalTravelDistance segmentsDepartureTimeEpochSeconds  \
2               7               1675.0            1650206700||1650229500   
3               7               1247.0            1650235800||1650246600   
4               7                762.0                        1650195000   
5               7                762.0                        1650200400   
6               7                762.0                        1650235800   

                            segmentsDepartureTimeRaw  \
2  2022-04-17T10:45:00.000-04:00||2022-04-17T17:0...   
3  2022-04-17T18:50:00.000-04:00||2022-04-17T21:5...   
4                      2022-04-17T07:30:00.000-04:00   
5                      2022-04-17T09:00:00.000-04:00   
6                      2022-04-17T18:50:00.000-04:00   

  segmentsArrivalTimeEpochSeconds  \
2          1650216180||1650238380   
3          1650244380||1650254280   
4                      1650202740   
5                      1650208500   
6                      1650244380   

                              segmentsArrivalTimeRaw  \
2  2022-04-17T13:23:00.000-04:00||2022-04-17T19:3...   
3  2022-04-17T21:13:00.000-04:00||2022-04-17T23:5...   
4                      2022-04-17T09:39:00.000-04:00   
5                      2022-04-17T11:15:00.000-04:00   
6                      2022-04-17T21:13:00.000-04:00   

  segmentsArrivalAirportCode segmentsDepartureAirportCode  \
2                   BOS||CLT                     ATL||BOS   
3                   JFK||DTW                     ATL||JFK   
4                        EWR                          ATL   
5                        JFK                          ATL   
6                        JFK                          ATL   

                segmentsAirlineName segmentsDurationInSeconds  \
2  JetBlue Airways||JetBlue Airways                9480||8880   
3  JetBlue Airways||JetBlue Airways                8580||7680   
4                   JetBlue Airways                      7740   
5                   JetBlue Airways                      8100   
6                   JetBlue Airways                      8580   

  segmentsDistance segmentsCabinCode  
2         947||728      coach||coach  
3         762||485      coach||coach  
4              762             coach  
5              762             coach  
6              762             coach

Create a seperate dataset for partnership flights

CREATE `days left` COLUMN

In [38]:
df.tail()

legId  searchDate  flightDate  \
6824435  6cf87e4163de23cc0ad1791685d37535  2022-10-05  2022-11-12   
6824436  a2c33f479416f3420866a35576d8e80e  2022-10-05  2022-11-12   
6824437  c33ada499ceb9ea377107976f23ed202  2022-10-05  2022-11-12   
6824438  ea55464ecfb2a65dc0b3c82139e3983c  2022-10-05  2022-11-12   
6824439  8ee566ea485df9c07d69ff7e392954b8  2022-10-05  2022-11-12   

        startingAirport destinationAirport fareBasisCode travelDuration  \
6824435             ORD                ATL      OI3QUEL1        PT13H4M   
6824436             ORD                DEN      PI4QUEL1        PT10H2M   
6824437             ORD                DFW      PI4QUEL1        PT7H43M   
6824438             ORD                LAX      PI4QUEL1       PT15H56M   
6824439             ORD                LAX      PI4QUEL1        PT9H19M   

         elapsedDays  isBasicEconomy  isRefundable  isNonStop  baseFare  \
6824435            1           False         False      False    156.28   
6824436            0           False         False      False    134.88   
6824437            0           False         False      False    106.97   
6824438            0           False         False      False    158.14   
6824439            0           False         False      False    190.69   

         totalFare  seatsRemaining  totalTravelDistance  \
6824435     197.20               7               1809.0   
6824436     168.60               7               2363.0   
6824437     138.59               7               2115.0   
6824438     199.20               7               3145.0   
6824439     228.59               7               3198.0   

        segmentsDepartureTimeEpochSeconds  \
6824435            1668301200||1668337500   
6824436            1668253500||1668274200   
6824437            1668289200||1668303000   
6824438            1668264420||1668299400   
6824439            1668253500||1668264540   

                                  segmentsDepartureTimeRaw  \
6824435  2022-11-12T19:00:00.000-06:00||2022-11-13T06:0...   
6824436  2022-11-12T05:45:00.000-06:00||2022-11-12T12:3...   
6824437  2022-11-12T15:40:00.000-06:00||2022-11-12T20:3...   
6824438  2022-11-12T08:47:00.000-06:00||2022-11-12T19:3...   
6824439  2022-11-12T05:45:00.000-06:00||2022-11-12T09:4...   

        segmentsArrivalTimeEpochSeconds  \
6824435          1668309480||1668348240   
6824436          1668261300||1668289620   
6824437          1668297360||1668316980   
6824438          1668272340||1668321780   
6824439          1668261300||1668287040   

                                    segmentsArrivalTimeRaw  \
6824435  2022-11-12T22:18:00.000-05:00||2022-11-13T09:0...   
6824436  2022-11-12T08:55:00.000-05:00||2022-11-12T14:4...   
6824437  2022-11-12T18:56:00.000-05:00||2022-11-12T23:2...   
6824438  2022-11-12T11:59:00.000-05:00||2022-11-12T22:4...   
6824439  2022-11-12T08:55:00.000-05:00||2022-11-12T13:0...   

        segmentsArrivalAirportCode segmentsDepartureAirportCode  \
6824435                   BOS||ATL                     ORD||BOS   
6824436                   JFK||DEN                     ORD||JFK   
6824437                   JFK||DFW                     ORD||JFK   
6824438                   JFK||ONT                     ORD||JFK   
6824439                   JFK||LAX                     ORD||JFK   

                      segmentsAirlineName segmentsDurationInSeconds  \
6824435  JetBlue Airways||JetBlue Airways               8280||10740   
6824436  JetBlue Airways||JetBlue Airways               7800||15420   
6824437  JetBlue Airways||JetBlue Airways               8160||13980   
6824438  JetBlue Airways||JetBlue Airways               7920||22380   
6824439  JetBlue Airways||JetBlue Airways               7800||22500   

        segmentsDistance segmentsCabinCode  
6824435         862||947      coach||coach  
6824436        731||1632      coach||coach  
6824437        731||1384      coach||coach  
6824438        731||2414      coach||coach  
6824439        731||2467      coa

In [39]:
print("Data types before conversion:")
print(df[['searchDate', 'flightDate']].dtypes)
print(df[['searchDate', 'flightDate']].head())

Data types before conversion:
searchDate    object
flightDate    object
dtype: object
   searchDate  flightDate
2  2022-04-16  2022-04-17
3  2022-04-16  2022-04-17
4  2022-04-16  2022-04-17
5  2022-04-16  2022-04-17
6  2022-04-16  2022-04-17


In [40]:
def days_left_col(df):
    # Convert 'searchDate' and 'flightDate' to datetime, handling potential errors
    df['searchDate'] = pd.to_datetime(df['searchDate'], format='%Y-%m-%d', errors='coerce')
    df['flightDate'] = pd.to_datetime(df['flightDate'], format='%Y-%m-%d', errors='coerce')
    
    # Ensure the columns are of datetime type
    df['searchDate'] = df['searchDate'].astype('datetime64[ns]')
    df['flightDate'] = df['flightDate'].astype('datetime64[ns]')
    
    print("\nData types after conversion:")
    print(df[['searchDate', 'flightDate']].dtypes)
    print(df[['searchDate', 'flightDate']].head())
    
    # Create the 'daysleft' column by calculating the difference
    df['daysleft'] = (df['flightDate'] - df['searchDate']).dt.days
    return df

days_left_col(df)

print('DataFrame:')
df.head()


Data types after conversion:
searchDate    datetime64[ns]
flightDate    datetime64[ns]
dtype: object
  searchDate flightDate
2 2022-04-16 2022-04-17
3 2022-04-16 2022-04-17
4 2022-04-16 2022-04-17
5 2022-04-16 2022-04-17
6 2022-04-16 2022-04-17
DataFrame:


legId searchDate flightDate startingAirport  \
2  a291a2b5ac19223c3407cfec58d1fc24 2022-04-16 2022-04-17             ATL   
3  d53bcf9e8b6523fa2de9420d7cb22196 2022-04-16 2022-04-17             ATL   
4  5bb138f39804320ece767e747608a50e 2022-04-16 2022-04-17             ATL   
5  8201f75d221c7aa5b78f3b7781166032 2022-04-16 2022-04-17             ATL   
6  a63591f2b4df39d809c4a45a3898e122 2022-04-16 2022-04-17             ATL   

  destinationAirport fareBasisCode travelDuration  elapsedDays  \
2                CLT      VH0AUEL1        PT8H48M            0   
3                DTW      RH0AUEL1         PT5H8M            0   
4                EWR      VC0QBEL1         PT2H9M            0   
5                JFK      RC0ABEL1        PT2H15M            0   
6                JFK      RC0ABEL1        PT2H23M            0   

   isBasicEconomy  isRefundable  isNonStop  baseFare  totalFare  \
2           False         False      False    482.79     542.60   
3           False         False      False    269.78     313.61   
4           False         False       True    171.16     198.60   
5           False         False       True    171.16     198.60   
6           False         False       True    171.16     198.60   

   seatsRemaining  totalTravelDistance segmentsDepartureTimeEpochSeconds  \
2               7               1675.0            1650206700||1650229500   
3               7               1247.0            1650235800||1650246600   
4               7                762.0                        1650195000   
5               7                762.0                        1650200400   
6               7                762.0                        1650235800   

                            segmentsDepartureTimeRaw  \
2  2022-04-17T10:45:00.000-04:00||2022-04-17T17:0...   
3  2022-04-17T18:50:00.000-04:00||2022-04-17T21:5...   
4                      2022-04-17T07:30:00.000-04:00   
5                      2022-04-17T09:00:00.000-04:00   
6                      2022-04-17T18:50:00.000-04:00   

  segmentsArrivalTimeEpochSeconds  \
2          1650216180||1650238380   
3          1650244380||1650254280   
4                      1650202740   
5                      1650208500   
6                      1650244380   

                              segmentsArrivalTimeRaw  \
2  2022-04-17T13:23:00.000-04:00||2022-04-17T19:3...   
3  2022-04-17T21:13:00.000-04:00||2022-04-17T23:5...   
4                      2022-04-17T09:39:00.000-04:00   
5                      2022-04-17T11:15:00.000-04:00   
6                      2022-04-17T21:13:00.000-04:00   

  segmentsArrivalAirportCode segmentsDepartureAirportCode  \
2                   BOS||CLT                     ATL||BOS   
3                   JFK||DTW                     ATL||JFK   
4                        EWR                          ATL   
5                        JFK                          ATL   
6                        JFK                          ATL   

                segmentsAirlineName segmentsDurationInSeconds  \
2  JetBlue Airways||JetBlue Airways                9480||8880   
3  JetBlue Airways||JetBlue Airways                8580||7680   
4                   JetBlue Airways                      7740   
5                   JetBlue Airways                      8100   
6                   JetBlue Airways                      8580   

  segmentsDistance segmentsCabinCode  daysleft  
2         947||728      coach||coach         1  
3         762||485      coach||coach         1  
4              762             coach         1  
5              762             coach         1  
6              762             coach         1

In [41]:
# Function to check if a flight is a partnership flight
partnership_flights=df[~df['segmentsAirlineName'].isin(['JetBlue Airways', 
                                  'JetBlue Airways||JetBlue Airways',
                                  'JetBlue Airways||JetBlue Airways||JetBlue Airways'])
 ]
jetblue_only_flights = df[df['segmentsAirlineName'].isin(['JetBlue Airways', 
                                  'JetBlue Airways||JetBlue Airways',
                                  'JetBlue Airways||JetBlue Airways||JetBlue Airways'])]
print('Partnership Flights dataset')
display(partnership_flights.head())

print(partnership_flights.groupby('segmentsAirlineName').size().sort_values(ascending=False))

print('\nJetblue-only Flights dataset')
display(jetblue_only_flights.head())

jetblue_only_flights.groupby('segmentsAirlineName').size().sort_values(ascending=False)

Partnership Flights dataset


legId searchDate flightDate startingAirport  \
2709   60b9af875d280cd59a1779446763fe65 2022-04-16 2022-04-23             ATL   
7015   60b9af875d280cd59a1779446763fe65 2022-04-17 2022-04-23             ATL   
9377   d431c81265674874a3f8f18a1f0ef8b1 2022-04-17 2022-04-27             JFK   
10676  e0542e15580d9e163701750792758fdb 2022-04-17 2022-04-29             JFK   
11275  dc066995b773eeb35b39dc378a7984c2 2022-04-17 2022-04-30             JFK   

      destinationAirport fareBasisCode travelDuration  elapsedDays  \
2709                 BOS      EH0JUEY5        PT7H35M            0   
7015                 BOS      EH0JUEY5        PT7H35M            0   
9377                 BOS           BIP        PT4H16M            0   
10676                BOS           KIP        PT4H16M            0   
11275                BOS           BIP        PT4H16M            0   

       isBasicEconomy  isRefundable  isNonStop  baseFare  totalFare  \
2709            False         False      False   1264.19    1378.10   
7015            False         False      False   1264.19    1378.10   
9377            False         False      False    198.40     232.38   
10676           False         False      False    263.52     302.38   
11275           False         False      False    198.40     232.38   

       seatsRemaining  totalTravelDistance  \
2709                1               1011.0   
7015                1               1011.0   
9377                1                472.0   
10676               1                472.0   
11275               1                472.0   

        segmentsDepartureTimeEpochSeconds  \
2709   1650718800||1650730200||1650743700   
7015   1650718800||1650730200||1650743700   
9377               1651078500||1651088700   
10676              1651251300||1651261500   
11275              1651337700||1651347900   

                                segmentsDepartureTimeRaw  \
2709   2022-04-23T09:00:00.000-04:00||2022-04-23T12:1...   
7015   2022-04-23T09:00:00.000-04:00||2022-04-23T12:1...   
9377   2022-04-27T12:55:00.000-04:00||2022-04-27T15:4...   
10676  2022-04-29T12:55:00.000-04:00||2022-04-29T15:4...   
11275  2022-04-30T12:55:00.000-04:00||2022-04-30T15:4...   

          segmentsArrivalTimeEpochSeconds  \
2709   1650726900||1650734580||1650746100   
7015   1650726900||1650734580||1650746100   
9377               1651085400||1651093860   
10676              1651258200||1651266660   
11275              1651344600||1651353060   

                                  segmentsArrivalTimeRaw  \
2709   2022-04-23T11:15:00.000-04:00||2022-04-23T13:2...   
7015   2022-04-23T11:15:00.000-04:00||2022-04-23T13:2...   
9377   2022-04-27T14:50:00.000-04:00||2022-04-27T17:1...   
10676  2022-04-29T14:50:00.000-04:00||2022-04-29T17:1...   
11275  2022-04-30T14:50:00.000-04:00||2022-04-30T17:1...   

      segmentsArrivalAirportCode segmentsDepartureAirportCode  \
2709               JFK||MVY||BOS                ATL||JFK||MVY   
7015               JFK||MVY||BOS                ATL||JFK||MVY   
9377                    SLK||BOS                     JFK||SLK   
10676                   SLK||BOS                     JFK||SLK   
11275                   SLK||BOS                     JFK||SLK   

                       segmentsAirlineName segmentsDurationInSeconds  \
2709   JetBlue Airways||Cape Air||Cape Air          8100||4380||2400   
7015   JetBlue Airways||Cape Air||Cape Air          8100||4380||2400   
9377             Cape Air||JetBlue Airways                6900||5160   
10676            Cape Air||JetBlue Airways                6900||5160   
11275            Cape Air||JetBlue Airways                6900||5160   

      segmentsDistance    segmentsCabinCode  daysleft  
2709      762||179||70  coach||coach||coach         7  
7015      762||179||70  coach||coach||coach         6  
9377          258||214         coach||coach        10  
10676         258||214         coach||coach        12  
11275         258||214         coach||coach        13

segmentsAirlineName
JetBlue Airways||Cape Air||Cape Air                 2142
Cape Air||JetBlue Airways                           1842
JetBlue Airways||Cape Air                           1353
JetBlue Airways||Cape Air||JetBlue Airways          1035
American Airlines||JetBlue Airways                   388
Cape Air||Cape Air||JetBlue Airways                  249
JetBlue Airways||JetBlue Airways||Cape Air           126
Cape Air||JetBlue Airways||JetBlue Airways           116
JetBlue Airways||Silver Airways||JetBlue Airways      80
JetBlue Airways||American Airlines                    13
Cape Air||Cape Air||Cape Air||JetBlue Airways          1
dtype: int64

Jetblue-only Flights dataset


legId searchDate flightDate startingAirport  \
2  a291a2b5ac19223c3407cfec58d1fc24 2022-04-16 2022-04-17             ATL   
3  d53bcf9e8b6523fa2de9420d7cb22196 2022-04-16 2022-04-17             ATL   
4  5bb138f39804320ece767e747608a50e 2022-04-16 2022-04-17             ATL   
5  8201f75d221c7aa5b78f3b7781166032 2022-04-16 2022-04-17             ATL   
6  a63591f2b4df39d809c4a45a3898e122 2022-04-16 2022-04-17             ATL   

  destinationAirport fareBasisCode travelDuration  elapsedDays  \
2                CLT      VH0AUEL1        PT8H48M            0   
3                DTW      RH0AUEL1         PT5H8M            0   
4                EWR      VC0QBEL1         PT2H9M            0   
5                JFK      RC0ABEL1        PT2H15M            0   
6                JFK      RC0ABEL1        PT2H23M            0   

   isBasicEconomy  isRefundable  isNonStop  baseFare  totalFare  \
2           False         False      False    482.79     542.60   
3           False         False      False    269.78     313.61   
4           False         False       True    171.16     198.60   
5           False         False       True    171.16     198.60   
6           False         False       True    171.16     198.60   

   seatsRemaining  totalTravelDistance segmentsDepartureTimeEpochSeconds  \
2               7               1675.0            1650206700||1650229500   
3               7               1247.0            1650235800||1650246600   
4               7                762.0                        1650195000   
5               7                762.0                        1650200400   
6               7                762.0                        1650235800   

                            segmentsDepartureTimeRaw  \
2  2022-04-17T10:45:00.000-04:00||2022-04-17T17:0...   
3  2022-04-17T18:50:00.000-04:00||2022-04-17T21:5...   
4                      2022-04-17T07:30:00.000-04:00   
5                      2022-04-17T09:00:00.000-04:00   
6                      2022-04-17T18:50:00.000-04:00   

  segmentsArrivalTimeEpochSeconds  \
2          1650216180||1650238380   
3          1650244380||1650254280   
4                      1650202740   
5                      1650208500   
6                      1650244380   

                              segmentsArrivalTimeRaw  \
2  2022-04-17T13:23:00.000-04:00||2022-04-17T19:3...   
3  2022-04-17T21:13:00.000-04:00||2022-04-17T23:5...   
4                      2022-04-17T09:39:00.000-04:00   
5                      2022-04-17T11:15:00.000-04:00   
6                      2022-04-17T21:13:00.000-04:00   

  segmentsArrivalAirportCode segmentsDepartureAirportCode  \
2                   BOS||CLT                     ATL||BOS   
3                   JFK||DTW                     ATL||JFK   
4                        EWR                          ATL   
5                        JFK                          ATL   
6                        JFK                          ATL   

                segmentsAirlineName segmentsDurationInSeconds  \
2  JetBlue Airways||JetBlue Airways                9480||8880   
3  JetBlue Airways||JetBlue Airways                8580||7680   
4                   JetBlue Airways                      7740   
5                   JetBlue Airways                      8100   
6                   JetBlue Airways                      8580   

  segmentsDistance segmentsCabinCode  daysleft  
2         947||728      coach||coach         1  
3         762||485      coach||coach         1  
4              762             coach         1  
5              762             coach         1  
6              762             coach         1

segmentsAirlineName
JetBlue Airways                                      3506867
JetBlue Airways||JetBlue Airways                     3092858
JetBlue Airways||JetBlue Airways||JetBlue Airways     108877
dtype: int64

Classifying `segmentsDepartureTimeRaw` and `segmentsArrivalTimeRaw` into categories

In [42]:
print('segmentDepartureTimeRaw')
display(df['segmentsDepartureTimeRaw'])


print('segmentArrivalTimeRaw')
display(df['segmentsArrivalTimeRaw'])

segmentDepartureTimeRaw


2          2022-04-17T10:45:00.000-04:00||2022-04-17T17:0...
3          2022-04-17T18:50:00.000-04:00||2022-04-17T21:5...
4                              2022-04-17T07:30:00.000-04:00
5                              2022-04-17T09:00:00.000-04:00
6                              2022-04-17T18:50:00.000-04:00
                                 ...                        
6824435    2022-11-12T19:00:00.000-06:00||2022-11-13T06:0...
6824436    2022-11-12T05:45:00.000-06:00||2022-11-12T12:3...
6824437    2022-11-12T15:40:00.000-06:00||2022-11-12T20:3...
6824438    2022-11-12T08:47:00.000-06:00||2022-11-12T19:3...
6824439    2022-11-12T05:45:00.000-06:00||2022-11-12T09:4...
Name: segmentsDepartureTimeRaw, Length: 6715947, dtype: object

segmentArrivalTimeRaw


2          2022-04-17T13:23:00.000-04:00||2022-04-17T19:3...
3          2022-04-17T21:13:00.000-04:00||2022-04-17T23:5...
4                              2022-04-17T09:39:00.000-04:00
5                              2022-04-17T11:15:00.000-04:00
6                              2022-04-17T21:13:00.000-04:00
                                 ...                        
6824435    2022-11-12T22:18:00.000-05:00||2022-11-13T09:0...
6824436    2022-11-12T08:55:00.000-05:00||2022-11-12T14:4...
6824437    2022-11-12T18:56:00.000-05:00||2022-11-12T23:2...
6824438    2022-11-12T11:59:00.000-05:00||2022-11-12T22:4...
6824439    2022-11-12T08:55:00.000-05:00||2022-11-12T13:0...
Name: segmentsArrivalTimeRaw, Length: 6715947, dtype: object

1. The times are in ISO 8601 format.
2. There might be multiple segments separated by "||" for flights with layovers.
3. extract the time from the first segment 

In [43]:
#function to convert segmentsDepartureTimeRaw' and 'segmentsArrivalTimeRaw' to time date
def extract_time(time_string):
    segments=time_string.split('||')
    first_segment=segments[0]
    #convert to datetime objects
    datetime=dt.fromisoformat(first_segment)
    return datetime.time() #return just time data

In [44]:
#TESTING THE METHOD

time_string = "2022-04-17T10:45:00.000-04:00||2022-04-17T13:23:00.000-04:00"
extracted_time = extract_time(time_string)
print(extracted_time)

10:45:00


In [45]:
#Use apply() for operations that need to be done row-wise or element-wise when a function takes individual elements or rows as inputs.
#Use direct function calls for vectorized operations or when the function operates on the entire DataFrame, as this is generally more efficient.
df['departureTime']=df['segmentsDepartureTimeRaw'].apply(extract_time)
df['ArrivalTime']=df['segmentsArrivalTimeRaw'].apply(extract_time)
# Print the DataFrame's name and show the tail to verify the new column
df.tail() # Display the last few rows to check the new column

legId searchDate flightDate  \
6824435  6cf87e4163de23cc0ad1791685d37535 2022-10-05 2022-11-12   
6824436  a2c33f479416f3420866a35576d8e80e 2022-10-05 2022-11-12   
6824437  c33ada499ceb9ea377107976f23ed202 2022-10-05 2022-11-12   
6824438  ea55464ecfb2a65dc0b3c82139e3983c 2022-10-05 2022-11-12   
6824439  8ee566ea485df9c07d69ff7e392954b8 2022-10-05 2022-11-12   

        startingAirport destinationAirport fareBasisCode travelDuration  \
6824435             ORD                ATL      OI3QUEL1        PT13H4M   
6824436             ORD                DEN      PI4QUEL1        PT10H2M   
6824437             ORD                DFW      PI4QUEL1        PT7H43M   
6824438             ORD                LAX      PI4QUEL1       PT15H56M   
6824439             ORD                LAX      PI4QUEL1        PT9H19M   

         elapsedDays  isBasicEconomy  isRefundable  isNonStop  baseFare  \
6824435            1           False         False      False    156.28   
6824436            0           False         False      False    134.88   
6824437            0           False         False      False    106.97   
6824438            0           False         False      False    158.14   
6824439            0           False         False      False    190.69   

         totalFare  seatsRemaining  totalTravelDistance  \
6824435     197.20               7               1809.0   
6824436     168.60               7               2363.0   
6824437     138.59               7               2115.0   
6824438     199.20               7               3145.0   
6824439     228.59               7               3198.0   

        segmentsDepartureTimeEpochSeconds  \
6824435            1668301200||1668337500   
6824436            1668253500||1668274200   
6824437            1668289200||1668303000   
6824438            1668264420||1668299400   
6824439            1668253500||1668264540   

                                  segmentsDepartureTimeRaw  \
6824435  2022-11-12T19:00:00.000-06:00||2022-11-13T06:0...   
6824436  2022-11-12T05:45:00.000-06:00||2022-11-12T12:3...   
6824437  2022-11-12T15:40:00.000-06:00||2022-11-12T20:3...   
6824438  2022-11-12T08:47:00.000-06:00||2022-11-12T19:3...   
6824439  2022-11-12T05:45:00.000-06:00||2022-11-12T09:4...   

        segmentsArrivalTimeEpochSeconds  \
6824435          1668309480||1668348240   
6824436          1668261300||1668289620   
6824437          1668297360||1668316980   
6824438          1668272340||1668321780   
6824439          1668261300||1668287040   

                                    segmentsArrivalTimeRaw  \
6824435  2022-11-12T22:18:00.000-05:00||2022-11-13T09:0...   
6824436  2022-11-12T08:55:00.000-05:00||2022-11-12T14:4...   
6824437  2022-11-12T18:56:00.000-05:00||2022-11-12T23:2...   
6824438  2022-11-12T11:59:00.000-05:00||2022-11-12T22:4...   
6824439  2022-11-12T08:55:00.000-05:00||2022-11-12T13:0...   

        segmentsArrivalAirportCode segmentsDepartureAirportCode  \
6824435                   BOS||ATL                     ORD||BOS   
6824436                   JFK||DEN                     ORD||JFK   
6824437                   JFK||DFW                     ORD||JFK   
6824438                   JFK||ONT                     ORD||JFK   
6824439                   JFK||LAX                     ORD||JFK   

                      segmentsAirlineName segmentsDurationInSeconds  \
6824435  JetBlue Airways||JetBlue Airways               8280||10740   
6824436  JetBlue Airways||JetBlue Airways               7800||15420   
6824437  JetBlue Airways||JetBlue Airways               8160||13980   
6824438  JetBlue Airways||JetBlue Airways               7920||22380   
6824439  JetBlue Airways||JetBlue Airways               7800||22500   

        segmentsDistance segmentsCabinCode  daysleft departureTime ArrivalTime  
6824435         862||947      coach||coach        38      19:00:00    22:18:00  
6824436        731||1632      coach||coach        38      05:45:00    08:55:00  
6824437        731||1384   

Classifying `departureTime` and `arrivalTime` into categories

Late Night: 23:00 - 02:00

Early Morning: 03:00 - 06:00

Morning: 07:00 - 10:00

Noon: 11:00 - 14:00

Afternoon: 15:00 - 18:00

Evening: 19:00 - 22:00

In [46]:
def time_classification(time):
    hour=time.hour #get the hour from the time object
    #ex: 06:30 -> 6
    if hour in range(3,7): #03:00 - 06:00
        return 'Early Morning'
    elif hour in range(7,11):
        return 'Morning'
    elif hour in range(11,15):
        return 'Noon'
    elif hour in range(15,19):
        return 'Afternoon'
    elif hour in range(19,23):
        return 'Evening'
    elif hour in [23,0,1,2]: # 23:00 - 02:00 (covering late night)
        return 'Late night'
    else:
        return 'Unknown'
df['departure_category'] = df['departureTime'].apply(time_classification)
df['arrival_category'] = df['ArrivalTime'].apply(time_classification)
df.tail()

legId searchDate flightDate  \
6824435  6cf87e4163de23cc0ad1791685d37535 2022-10-05 2022-11-12   
6824436  a2c33f479416f3420866a35576d8e80e 2022-10-05 2022-11-12   
6824437  c33ada499ceb9ea377107976f23ed202 2022-10-05 2022-11-12   
6824438  ea55464ecfb2a65dc0b3c82139e3983c 2022-10-05 2022-11-12   
6824439  8ee566ea485df9c07d69ff7e392954b8 2022-10-05 2022-11-12   

        startingAirport destinationAirport fareBasisCode travelDuration  \
6824435             ORD                ATL      OI3QUEL1        PT13H4M   
6824436             ORD                DEN      PI4QUEL1        PT10H2M   
6824437             ORD                DFW      PI4QUEL1        PT7H43M   
6824438             ORD                LAX      PI4QUEL1       PT15H56M   
6824439             ORD                LAX      PI4QUEL1        PT9H19M   

         elapsedDays  isBasicEconomy  isRefundable  isNonStop  baseFare  \
6824435            1           False         False      False    156.28   
6824436            0           False         False      False    134.88   
6824437            0           False         False      False    106.97   
6824438            0           False         False      False    158.14   
6824439            0           False         False      False    190.69   

         totalFare  seatsRemaining  totalTravelDistance  \
6824435     197.20               7               1809.0   
6824436     168.60               7               2363.0   
6824437     138.59               7               2115.0   
6824438     199.20               7               3145.0   
6824439     228.59               7               3198.0   

        segmentsDepartureTimeEpochSeconds  \
6824435            1668301200||1668337500   
6824436            1668253500||1668274200   
6824437            1668289200||1668303000   
6824438            1668264420||1668299400   
6824439            1668253500||1668264540   

                                  segmentsDepartureTimeRaw  \
6824435  2022-11-12T19:00:00.000-06:00||2022-11-13T06:0...   
6824436  2022-11-12T05:45:00.000-06:00||2022-11-12T12:3...   
6824437  2022-11-12T15:40:00.000-06:00||2022-11-12T20:3...   
6824438  2022-11-12T08:47:00.000-06:00||2022-11-12T19:3...   
6824439  2022-11-12T05:45:00.000-06:00||2022-11-12T09:4...   

        segmentsArrivalTimeEpochSeconds  \
6824435          1668309480||1668348240   
6824436          1668261300||1668289620   
6824437          1668297360||1668316980   
6824438          1668272340||1668321780   
6824439          1668261300||1668287040   

                                    segmentsArrivalTimeRaw  \
6824435  2022-11-12T22:18:00.000-05:00||2022-11-13T09:0...   
6824436  2022-11-12T08:55:00.000-05:00||2022-11-12T14:4...   
6824437  2022-11-12T18:56:00.000-05:00||2022-11-12T23:2...   
6824438  2022-11-12T11:59:00.000-05:00||2022-11-12T22:4...   
6824439  2022-11-12T08:55:00.000-05:00||2022-11-12T13:0...   

        segmentsArrivalAirportCode segmentsDepartureAirportCode  \
6824435                   BOS||ATL                     ORD||BOS   
6824436                   JFK||DEN                     ORD||JFK   
6824437                   JFK||DFW                     ORD||JFK   
6824438                   JFK||ONT                     ORD||JFK   
6824439                   JFK||LAX                     ORD||JFK   

                      segmentsAirlineName segmentsDurationInSeconds  \
6824435  JetBlue Airways||JetBlue Airways               8280||10740   
6824436  JetBlue Airways||JetBlue Airways               7800||15420   
6824437  JetBlue Airways||JetBlue Airways               8160||13980   
6824438  JetBlue Airways||JetBlue Airways               7920||22380   
6824439  JetBlue Airways||JetBlue Airways               7800||22500   

        segmentsDistance segmentsCabinCode  daysleft departureTime  \
6824435         862||947      coach||coach        38      19:00:00   
6824436        731||1632      coach||coach        38      05:45:00   
6824437        731||1384      coach||coach        38      15

In [47]:
df.groupby(df['departure_category']).size().sort_values(ascending=False)

departure_category
Morning          1964440
Noon             1801835
Afternoon        1185638
Evening           833587
Early Morning     603226
Late night        327221
dtype: int64

NO `Unknown` VALUES IN BOTH `departure_category` AND `arrival_category`

In [48]:
df.groupby(df['arrival_category']).size().sort_values(ascending=False)

arrival_category
Noon             1800660
Morning          1608593
Afternoon        1418000
Evening          1245487
Late night        358480
Early Morning     284727
dtype: int64

#PLOTTING How is the price affected when tickets are bought in just 1 or 2 days before departure?

In [49]:
df.describe()

searchDate                     flightDate  \
count                        6715947                        6715947   
mean   2022-07-14 22:44:14.083959296  2022-08-11 07:16:07.442236160   
min              2022-04-16 00:00:00            2022-04-17 00:00:00   
25%              2022-06-07 00:00:00            2022-07-04 00:00:00   
50%              2022-07-17 00:00:00            2022-08-16 00:00:00   
75%              2022-08-25 00:00:00            2022-09-20 00:00:00   
max              2022-10-05 00:00:00            2022-11-19 00:00:00   
std                              NaN                            NaN   

        elapsedDays      baseFare     totalFare  seatsRemaining  \
count  6.715947e+06  6.715947e+06  6.715947e+06    6.715947e+06   
mean   1.443850e-01  2.312117e+02  2.682336e+02    6.701785e+00   
min    0.000000e+00  2.233000e+01  3.861000e+01    1.000000e+00   
25%    0.000000e+00  1.339500e+02  1.586000e+02    7.000000e+00   
50%    0.000000e+00  1.934900e+02  2.272000e+02    7.000000e+00   
75%    0.000000e+00  2.888400e+02  3.296000e+02    7.000000e+00   
max    1.000000e+00  3.770230e+03  4.076600e+03    1.000000e+01   
std    3.514797e-01  1.651109e+02  1.786933e+02    1.132687e+00   

       totalTravelDistance      daysleft  
count         6.715947e+06  6.715947e+06  
mean          1.530711e+03  2.735548e+01  
min           1.210000e+02  1.000000e+00  
25%           7.200000e+02  1.400000e+01  
50%           1.260000e+03  2.700000e+01  
75%           2.461000e+03  4.000000e+01  
max           3.958000e+03  6.000000e+01  
std           9.789898e+02  1.613734e+01